TWTValidation/JSON Validator/Schema Validators/TWTJSONSchemaArrayValidator.m

```diff
@@ -0,0 +1,155 @@
+    other.requiresUniqueItems == self.requiresUniqueItems && [other.itemValidator isEqual:self.itemValidator] &&
+    [other.indexedItemValidators isEqual:self.indexedItemValidators] && [other.additionalItemsValidator isEqual:self.additionalItemsValidator];
+
+}
+
+
+- (BOOL)validateValue:(id)value error:(out NSError *__autoreleasing *)outError
+{
+    if (![super validateValue:value error:outError]) {
+        return NO;
+    } else if (![value isKindOfClass:[NSArray class]]) {
+        if (outError) {
+            //            *outError = [NSError twt_validationErrorWithCode:TWTValidationErrorCodeValueHasIncorrectClass
+            //                                            failingValidator:self
+            //                                                       value:value
+            //                                        localizedDescription:TWTLocalizedString(@"TWTJSONSchemaArrayValidator.notArrayError")];
+        }
+        return NO;
+    }
+
+    BOOL countValidated = YES;
+    BOOL uniqueItemsValidated = YES;
+    BOOL itemsValidated = YES;
+    BOOL additionalItemsValidated = YES;
+
+    NSError *countError = nil;
+    NSError *uniqueItemsError = nil;
+    NSMutableArray *itemsErrors = outError ? [[NSMutableArray alloc] init] : nil;
+    NSMutableArray *additionalItemsErrors = outError ? [[NSMutableArray alloc] init] : nil;
+
+    if (self.maximumItemCount || self.minimumItemCount) {
+        TWTValidator *countValidator = [[TWTNumberValidator alloc] initWithMinimum:self.minimumItemCount maximum:self.maximumItemCount];
+        countValidated = [countValidator validateValue:@([value count]) error:&countError];
+    }
+
+    if (self.requiresUniqueItems) {
+        NSCountedSet *itemSet = [[NSCountedSet alloc] initWithArray:value];
+        NSSet *repeats = [itemSet objectsPassingTest:^BOOL(id object, BOOL *stop) {
+            return [itemSet countForObject:object] > 1;
+        }];
+        if (repeats.count > 0) {
+            uniqueItemsValidated = NO;
+            //            uniqueItemsError =
+        }
+    }
+
+    NSError *error = nil;
+
+
+    if (self.itemValidator) {
+        for (id item in value) {
+            error = nil;
+            if (![self.itemValidator validateValue:item error:outError ? &error : NULL]) {
+                itemsValidated = NO;
+                //                    itemsErrors addObject:
+            }
+        }
+    } else if (self.indexedItemValidators) {
+        NSUInteger index = 0;
+        for (id item in value) {
+            error = nil;
+
+            if (index < self.indexedItemValidators.count) {
+                if (![self.indexedItemValidators[index] validateValue:item error:outError ? &error : NULL]) {
+                    itemsValidated = NO;
+                    //                        add error
+                }
+            } else {
+                if (![self.additionalItemsValidator validateValue:item error:outError ? &error : NULL]) {
```

Comment:
Yes it's an always passing validator. A default `YES` is set when the array node is initialized, (and can then be changed during parsing) so it will always produce some kind of validator.